metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,458 @@
+      if (duplicatedKeys) {
+        None
+      } else {
+        Some(new ClusterConfig[K](connectionsToAcquire, allowedIncoming) {})
+      }
+    }
+  }
+
+  /** Connection manager for static topology cluster. It starts 3 concurrent backgrounds processes:
+    * 1. Calling process - tries to connect to remote nodes specified in cluster config. In case of failure, retries with
+    *    exponential backoff.
+    * 2. Server process - reads incoming connections from server socket. Validates that incoming connections is from known
+    *    remote peer specified in cluster config.
+    * 3. Message reading process - receives connections from both, Calling and Server processes, and for each connections
+    *    start concurrent process reading messages from those connections. In case of some error on connections, it closes
+    *    connection. In case of discovering that one of outgoing connections failed, it request Calling process to establish
+    *    connection once again.
+    *
+    * @param encryptedConnectionsProvider component which makes it possible to receive and acquire encrypted connections
+    * @param clusterConfig static cluster topology configuration
+    * @param retryConfig retry configuration for outgoing connections (incoming connections are not retried)
+    */
+  def apply[F[_]: Concurrent: TaskLift: TaskLike: Timer, K: Codec, M: Codec](
+      encryptedConnectionsProvider: EncryptedConnectionProvider[F, K, M],
+      clusterConfig: ClusterConfig[K],
+      retryConfig: RetryConfig
+  )(implicit
+      cs: ContextShift[F]
+  ): Resource[F, RemoteConnectionManager[F, K, M]] = {
+    for {
+      acquiredConnections <- Resource.liftF(ConnectionsRegister.empty[F, K, M])
+      connectionsToAcquireQueue <- Resource.liftF(
+        ConcurrentQueue.unbounded[F, OutGoingConnectionRequest[K]]()
+      )
+      _ <- Resource.liftF(
+        connectionsToAcquireQueue.offerMany(
+          clusterConfig.connectionsToAcquire.map { case (key, address) =>
+            OutGoingConnectionRequest.initial(key, address)
+          }
+        )
+      )
+      connectionQueue <- Resource.liftF(
+        ConcurrentQueue.unbounded[F, EncryptedConnection[F, K, M]]()
+      )
+      messageQueue <- Resource.liftF(
+        ConcurrentQueue.unbounded[F, MessageReceived[K, M]]()
+      )
+
+      _ <- acquireConnections(
+        encryptedConnectionsProvider,
+        connectionsToAcquireQueue,
+        acquiredConnections,
+        connectionQueue,
+        retryConfig
+      ).background
+      _ <- handleServerConnections(
+        encryptedConnectionsProvider,
+        connectionQueue,
+        acquiredConnections,
+        clusterConfig
+      ).background
+
+      _ <- handleConnections(
+        connectionQueue,
+        acquiredConnections,
+        connectionsToAcquireQueue,
+        messageQueue,
+        clusterConfig
+      ).background
+    } yield new RemoteConnectionManager[F, K, M](
```

Comment:
Perhaps it would make sense to wait until the cluster can be considered connected, for example at least 1 attempt has been made to connect to everybody, before we start trying to fire off messages?

Reply:
I would leave the down stream responsible for this decision. Currently anyone can start `RemoteConnectionManager` and wait until `acquiredConnections` returns some minimal number of connected connections.

Reply:
That's true; what I meant wasn't that we should wait until we have a certain number of connections, but rather that it has _attempted_ to make a connection. But you're right, it's more complicated than it's worth, and since we know we need a certain number of connections for quorum, we might as well wait until we have at least that many peers.